File: backend/src/services/cluster-upgrade.service.ts
import logger from "../logger/logger";
import { IElasticNode } from "../models/cluster-node.model";
import { getClusterInfoById } from "./cluster-info.service";
import { ansibleInventoryService } from "./ansible-inventory.service";
import { ansibleRunnerService } from "./ansible-runner.service";
import { ClusterType, NodeStatus } from "../enums";
import { randomUUID } from "crypto";
import { NotificationEventType, notificationService, NotificationType } from "./notification.service";
import { clusterUpgradeJobService } from "./cluster-upgrade-job.service";
import { clusterNodeService } from "./cluster-node.service";

class ClusterUpgradeService {
	async triggerElasticNodeUpgrade(nodeId: string, clusterId: string) {
		try {
			const node = await clusterNodeService.getElasticNodeById(nodeId);
			if (!node) {
				return false;
			}
			const clusterInfo = await getClusterInfoById(clusterId);
			const clusterUpgradeJob = await clusterUpgradeJobService.getActiveClusterUpgradeJobByClusterId(clusterId);
			const pathToKey = clusterInfo.pathToKey ? clusterInfo.pathToKey : "";

			const inventoryPath = await ansibleInventoryService.createInventoryForNode({
				node: node,
				pathToKey,
				sshUser: clusterInfo.sshUser,
			});

			const playbookRunId = randomUUID();

			ansibleRunnerService
				.runPlaybook({
					playbookPath: "playbooks/main.yml",
					inventoryPath: inventoryPath,
					variables: {
						elk_version: clusterUpgradeJob.targetVersion,
						es_username: clusterInfo.elastic.username,
						es_password: clusterInfo.elastic.password,
						es_api_key: clusterInfo.elastic.apiKey,
						elasticsearch_uri: clusterInfo.elastic.url,
						cluster_type: ClusterType.ELASTIC,
						playbook_run_id: playbookRunId,
						playbook_run_type: "UPGRADE",
						current_version: clusterUpgradeJob.currentVersion,
					},
				})
				.then(() => {
					notificationService.sendNotification({
						type: NotificationEventType.NOTIFICATION,
						title: "Upgrade Successful",
						message: "Node has been successfully upgraded to the target version.",
						notificationType: NotificationType.SUCCESS,
						clusterId: clusterId,
					});
				})
				.catch(() => {
					notificationService.sendNotification({
						type: NotificationEventType.NOTIFICATION,
						title: "Upgrade Failed",
						message: "An error occurred while upgrading the node. Please check the logs for more details.",
						notificationType: NotificationType.ERROR,
					});
				});

			return new Promise((resolve) => resolve(true));
		} catch (error) {
			logger.error(`Error performing upgrade for node with id ${nodeId}`);
			return false;
		}
	}

	async triggerElasticNodesUpgrade(nodes: IElasticNode[], clusterId: string) {
		try {
			const clusterInfo = await getClusterInfoById(clusterId);
			const clusterUpgradeJob = await clusterUpgradeJobService.getActiveClusterUpgradeJobByClusterId(clusterId);
			const pathToKey = clusterInfo.pathToKey ? clusterInfo.pathToKey : "";

			const inventoryPath = await ansibleInventoryService.createInventory({
				pathToKey,
				sshUser: clusterInfo.sshUser,
				nodes: nodes,
			});

			const playbookRunId = randomUUID();

			ansibleRunnerService
				.runPlaybook({
					playbookPath: "playbooks/main.yml",
					inventoryPath: inventoryPath,
					variables: {
						elk_version: clusterUpgradeJob.targetVersion,
						es_username: clusterInfo.elastic.username,
						es_password: clusterInfo.elastic.password,
						es_api_key: clusterInfo.elastic.apiKey,
						elasticsearch_uri: clusterInfo.elastic.url,
						cluster_type: ClusterType.ELASTIC,
						playbook_run_id: playbookRunId,
						playbook_run_type: "UPGRADE",
						current_version: clusterUpgradeJob.currentVersion,
					},
				})
				.then(() => {
					notificationService.sendNotification({
						type: NotificationEventType.NOTIFICATION,
						title: "Upgrade Successful",
						message: "All nodes have been successfully upgraded to the target version.",
						notificationType: NotificationType.SUCCESS,
						clusterId: clusterId,
					});
				})
				.catch(() => {
					notificationService.sendNotification({
						type: NotificationEventType.NOTIFICATION,
						title: "Upgrade Failed",
						message: "An error occurred while upgrading the nodes. Please check the logs for more details.",
						notificationType: NotificationType.ERROR,
					});
				});
		} catch (error: any) {
			logger.error(`Error performing upgrade for nodes:  ${nodes} because of ${error.message}`);
			throw new Error(`Error performing upgrade for nodes:  ${nodes}`);
		}
	}

	async triggerKibanaNodeUpgrade(nodeId: string, clusterId: string) {
		try {
			const node = await clusterNodeService.getKibanaNodeById(nodeId);
			if (!node) {
				logger.error(`Kibana node not found for [nodeId:${nodeId}]`);
				return;
			}
			const clusterInfo = await getClusterInfoById(clusterId);
			const clusterUpgradeJob = await clusterUpgradeJobService.getActiveClusterUpgradeJobByClusterId(clusterId);
			const pathToKey = clusterInfo.pathToKey ? clusterInfo.pathToKey : ""; //Should be stored in clusterInfo
			const inventoryPath = await ansibleInventoryService.createInventoryForNode({
				pathToKey,
				sshUser: clusterInfo.sshUser,
				node: node,
			});

			const playbookRunId = randomUUID();

			ansibleRunnerService
				.runPlaybook({
					playbookPath: "playbooks/main.yml",
					inventoryPath: inventoryPath,
					variables: {
						elk_version: clusterUpgradeJob.targetVersion,
						es_username: clusterInfo.elastic.username,
						es_password: clusterInfo.elastic.password,
						es_api_key: clusterInfo.elastic.apiKey,
						elasticsearch_uri: clusterInfo.elastic.url,
						cluster_type: ClusterType.KIBANA,
						playbook_run_id: playbookRunId,
						playbook_run_type: "UPGRADE",
						current_version: clusterUpgradeJob.currentVersion,
					},
				})
				.then(async () => {
					await clusterNodeService.updateNodesPartially(
						{ nodeId: nodeId },
						{ status: NodeStatus.UPGRADED, progress: 100, version: clusterUpgradeJob.targetVersion }
					);
					notificationService.sendNotification({
						type: NotificationEventType.NOTIFICATION,
						title: "Upgrade Successful",
						message: "Kibana node has been successfully upgraded to the target version.",
						notificationType: NotificationType.SUCCESS,
						clusterId: clusterId,
					});
				})
				.catch(async () => {
					await clusterNodeService.updateNodesPartially({ nodeId: nodeId }, { status: NodeStatus.FAILED });
					notificationService.sendNotification({
						type: NotificationEventType.NOTIFICATION,
						title: "Upgrade Failed",
						message:
							"An error occurred while upgrading the kibana node. Please check the logs for more details.",
						notificationType: NotificationType.ERROR,
						clusterId: clusterId,
					});
				});
		} catch (error) {
			logger.error(`Error performing upgrade for node with id ${nodeId}`);
		}
	}
}

export const clusterUpgradeService = new ClusterUpgradeService();

File: server/src/main/java/co/hyperflex/services/ClusterUpgradeService.java
package co.hyperflex.services;

import co.hyperflex.clients.elastic.ElasticClient;
import co.hyperflex.clients.elastic.ElasticsearchClientProvider;
import co.hyperflex.clients.elastic.dto.GetElasticsearchSnapshotResponse;
import co.hyperflex.clients.kibana.KibanaClient;
import co.hyperflex.clients.kibana.KibanaClientProvider;
import co.hyperflex.common.exceptions.BadRequestException;
import co.hyperflex.core.models.enums.ClusterNodeType;
import co.hyperflex.core.models.enums.ClusterUpgradeStatus;
import co.hyperflex.core.models.enums.NodeUpgradeStatus;
import co.hyperflex.core.services.clusters.ClusterService;
import co.hyperflex.core.services.clusters.dtos.GetClusterNodeResponse;
import co.hyperflex.core.services.clusters.dtos.GetClusterResponse;
import co.hyperflex.core.services.clusters.dtos.GetElasticCloudClusterResponse;
import co.hyperflex.core.services.clusters.lock.ClusterLockService;
import co.hyperflex.dtos.ClusterInfoResponse;
import co.hyperflex.dtos.upgrades.ClusterNodeUpgradeRequest;
import co.hyperflex.dtos.upgrades.ClusterNodeUpgradeResponse;
import co.hyperflex.dtos.upgrades.ClusterUpgradeResponse;
import co.hyperflex.entities.cluster.ClusterEntity;
import co.hyperflex.entities.cluster.ClusterNodeEntity;
import co.hyperflex.entities.cluster.SelfManagedClusterEntity;
import co.hyperflex.entities.precheck.PrecheckStatus;
import co.hyperflex.entities.upgrade.ClusterUpgradeJobEntity;
import co.hyperflex.entities.upgrade.UpgradeLogEntity;
import co.hyperflex.prechecks.scheduler.PrecheckSchedulerService;
import co.hyperflex.repositories.ClusterNodeRepository;
import co.hyperflex.repositories.ClusterRepository;
import co.hyperflex.services.notifications.GeneralNotificationEvent;
import co.hyperflex.services.notifications.NotificationService;
import co.hyperflex.services.notifications.NotificationType;
import co.hyperflex.services.notifications.UpgradeProgressChangeEvent;
import co.hyperflex.upgrader.planner.UpgradePlanBuilder;
import co.hyperflex.upgrader.tasks.Configuration;
import co.hyperflex.upgrader.tasks.Context;
import co.hyperflex.upgrader.tasks.Task;
import co.hyperflex.upgrader.tasks.TaskResult;
import co.hyperflex.utils.VersionUtils;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;
import org.springframework.stereotype.Service;

@Service
public class ClusterUpgradeService {
  private static final Logger log = LoggerFactory.getLogger(ClusterUpgradeService.class);
  private final ElasticsearchClientProvider elasticsearchClientProvider;
  private final ClusterNodeRepository clusterNodeRepository;
  private final ClusterService clusterService;
  private final ClusterRepository clusterRepository;
  private final KibanaClientProvider kibanaClientProvider;
  private final ClusterUpgradeJobService clusterUpgradeJobService;
  private final NotificationService notificationService;
  private final PrecheckSchedulerService precheckSchedulerService;
  private final DeprecationService deprecationService;
  private final ExecutorService executorService = Executors.newFixedThreadPool(1);
  private final PrecheckRunService precheckRunService;
  private final ClusterLockService clusterLockService;

  public ClusterUpgradeService(ElasticsearchClientProvider elasticsearchClientProvider, ClusterNodeRepository clusterNodeRepository,
                               ClusterService clusterService, ClusterRepository clusterRepository,
                               KibanaClientProvider kibanaClientProvider, ClusterUpgradeJobService clusterUpgradeJobService,
                               NotificationService notificationService, PrecheckSchedulerService precheckSchedulerService,
                               DeprecationService deprecationService, PrecheckRunService precheckRunService,
                               ClusterLockService clusterLockService) {
    this.elasticsearchClientProvider = elasticsearchClientProvider;
    this.clusterNodeRepository = clusterNodeRepository;
    this.clusterService = clusterService;
    this.clusterRepository = clusterRepository;
    this.kibanaClientProvider = kibanaClientProvider;
    this.clusterUpgradeJobService = clusterUpgradeJobService;
    this.notificationService = notificationService;
    this.precheckSchedulerService = precheckSchedulerService;
    this.deprecationService = deprecationService;
    this.precheckRunService = precheckRunService;
    this.clusterLockService = clusterLockService;
  }

  public ClusterNodeUpgradeResponse upgradeNode(ClusterNodeUpgradeRequest request) {

    ClusterEntity cluster = clusterRepository.findById(request.clusterId()).orElseThrow();
    if (cluster instanceof SelfManagedClusterEntity selfManagedCluster) {
      ClusterUpgradeJobEntity clusterUpgradeJob = clusterUpgradeJobService.getActiveJobByClusterId(request.clusterId());
      ClusterNodeEntity clusterNode = clusterNodeRepository.findById(request.nodeId()).orElseThrow();
      upgradeNodes(selfManagedCluster, List.of(clusterNode), clusterUpgradeJob.getId());
    } else {
      throw new BadRequestException("Upgrade not supported for cluster");
    }
    return new ClusterNodeUpgradeResponse("Node upgrade started");
  }

  public ClusterUpgradeResponse upgrade(String clusterId, ClusterNodeType nodeType) {
    ClusterEntity cluster = clusterRepository.findById(clusterId).orElseThrow();
    if (cluster instanceof SelfManagedClusterEntity selfManagedCluster) {
      ClusterUpgradeJobEntity clusterUpgradeJob = clusterUpgradeJobService.getActiveJobByClusterId(clusterId);
      List<ClusterNodeEntity> clusterNodes =
          clusterNodeRepository.findByClusterId(clusterId).stream().filter(node -> node.getType() == nodeType)
              .sorted(Comparator.comparingInt(ClusterNodeEntity::getRank)).toList();

      upgradeNodes(selfManagedCluster, clusterNodes, clusterUpgradeJob.getId());
    } else {
      throw new BadRequestException("Upgrade not supported for cluster");
    }
    return new ClusterUpgradeResponse("Cluster upgrade started");
  }

  public ClusterInfoResponse upgradeInfo(String clusterId) {
    ClusterUpgradeJobEntity activeUpgradeJob = null;
    try {
      activeUpgradeJob = clusterUpgradeJobService.getActiveJobByClusterId(clusterId);
    } catch (Exception e) {
      log.error("Failed to retrieve active job for clusterId: {}", clusterId, e);
    }
    try {
      ElasticClient client = elasticsearchClientProvider.getClient(clusterId);
      KibanaClient kibanaClient = kibanaClientProvider.getClient(clusterId);
      GetClusterResponse cluster = clusterService.getClusterById(clusterId);
      PrecheckStatus precheckStatus = null;
      if (activeUpgradeJob != null) {
        boolean isClusterUpgrading = activeUpgradeJob.getStatus() == ClusterUpgradeStatus.UPGRADING
            || activeUpgradeJob.getStatus() == ClusterUpgradeStatus.PARTIALLY_UPDATED
            || activeUpgradeJob.getStatus() == ClusterUpgradeStatus.FAILED;
        if (isClusterUpgrading) {
          precheckStatus = PrecheckStatus.COMPLETED;
        } else {
          boolean precheckExists = precheckRunService.precheckExistsForJob(activeUpgradeJob.getId());
          if (!precheckExists) {
            precheckSchedulerService.schedule(clusterId);
            precheckStatus = PrecheckStatus.RUNNING;
          } else {
            precheckStatus = precheckRunService.getStatusByUpgradeJobId(activeUpgradeJob.getId());
          }
        }
      } else {
        precheckStatus = PrecheckStatus.COMPLETED;
      }

      List<GetElasticsearchSnapshotResponse> snapshots = client.getValidSnapshots();

      ClusterInfoResponse.DeprecationCounts kibanaDeprecationCounts = deprecationService.getKibanaDeprecationCounts(clusterId);
      ClusterInfoResponse.DeprecationCounts elasticDeprecationCounts = deprecationService.getElasticDeprecationCounts(clusterId);

      boolean isClusterUpgraded = activeUpgradeJob != null && activeUpgradeJob.getStatus() == ClusterUpgradeStatus.UPDATED;
      // Evaluate upgrade status
      boolean isESUpgraded = isClusterUpgraded || clusterService.isNodesUpgraded(clusterId, ClusterNodeType.ELASTIC);
      boolean isKibanaUpgraded = isClusterUpgraded || clusterService.isNodesUpgraded(clusterId, ClusterNodeType.KIBANA);

      boolean isElasticUpgradable = !isESUpgraded;
      boolean isKibanaUpgradable = !isKibanaUpgraded && isESUpgraded;

      ClusterInfoResponse.Elastic elastic = new ClusterInfoResponse.Elastic(isElasticUpgradable, elasticDeprecationCounts,
          new ClusterInfoResponse.Elastic.SnapshotWrapper(snapshots.isEmpty() ? null : snapshots.getFirst(),
              kibanaClient.getSnapshotCreationPageUrl()));

      ClusterInfoResponse.Kibana kibana = new ClusterInfoResponse.Kibana(isKibanaUpgradable, kibanaDeprecationCounts);

      ClusterInfoResponse.Precheck precheck = new ClusterInfoResponse.Precheck(precheckStatus);
      String deploymentId = cluster instanceof GetElasticCloudClusterResponse elasticCloud ? elasticCloud.getDeploymentId() : null;
      return new ClusterInfoResponse(elastic, kibana, precheck, deploymentId);
    } catch (Exception e) {
      log.error("Failed to get upgrade info for clusterId: {}", clusterId, e);
      throw new RuntimeException(e);
    }
  }

  private void upgradeNodes(SelfManagedClusterEntity cluster, List<ClusterNodeEntity> nodes, String clusterUpgradeJobId) {

    executorService.submit(() -> {
      try {
        clusterLockService.lock(cluster.getId());

        clusterUpgradeJobService.setJobStatus(clusterUpgradeJobId, ClusterUpgradeStatus.UPGRADING);
        MDC.put(UpgradeLogEntity.CLUSTER_UPGRADE_JOB_ID, clusterUpgradeJobId);

        ElasticClient elasticClient = elasticsearchClientProvider.getClient(cluster.getId());
        KibanaClient kibanaClient = kibanaClientProvider.getClient(cluster);

        final String targetVersion = clusterUpgradeJobService.getUpgradeJobById(clusterUpgradeJobId).getTargetVersion();

        for (ClusterNodeEntity node : nodes.stream().sorted(Comparator.comparingInt(ClusterNodeEntity::getRank)).toList()) {

          if (clusterUpgradeJobService.getUpgradeJobById(clusterUpgradeJobId).isStop()) {
            clusterUpgradeJobService.setJobStatus(clusterUpgradeJobId, ClusterUpgradeStatus.STOPPED);
            notifyClusterUpgradedStopped(cluster);
            break;
          }

          MDC.put(UpgradeLogEntity.NODE_ID, node.getId());
          if (NodeUpgradeStatus.UPGRADED == node.getStatus()) {
            log.info("Skipping node with [NodeId: {}] as its already updated", node.getId());
            continue;
          } else if (VersionUtils.isVersionGte(node.getVersion(), targetVersion)) {
            log.info("Skipping node with [NodeId: {}] as its already on target version", node.getId());
            node.setStatus(NodeUpgradeStatus.UPGRADED);
            updateNodeProgress(node, 100);
            continue;
          }

          Configuration config =
              new Configuration(9300, 9200, cluster.getSshInfo().username(), cluster.getSshInfo().keyPath(), targetVersion);
          Context context = new Context(node, config, log, elasticClient, kibanaClient);

          UpgradePlanBuilder upgradePlanBuilder = new UpgradePlanBuilder();
          List<Task> tasks = upgradePlanBuilder.buildPlanFor(node);

          int checkPoint = clusterUpgradeJobService.getCheckPoint(clusterUpgradeJobId, node.getId());

          node.setStatus(NodeUpgradeStatus.UPGRADING);
          updateNodeProgress(node, 0);

          int index = 0;
          for (Task task : tasks) {
            if (index++ < checkPoint) {
              log.info("Skipping task [name: {}] for node [id: {}] — already upgraded.", task.getName(), node.getId());
              continue;
            }
            try {
              log.info("Starting task [name: {}] for node [ip: {}] [progress: {}%]", task.getName(), node.getIp(),
                  (index * 100) / tasks.size());

              TaskResult result = task.run(context);


              log.info("Task [name: {}] completed for node [ip: {}] [success: {}] [result: {}]", task.getName(), node.getIp(),
                  result.isSuccess(), result.getMessage());

              if (!result.isSuccess()) {
                log.error("Task [name: {}] failed for node [ip: {}] — {}", task.getName(), node.getIp(), result.getMessage());
                throw new RuntimeException(result.getMessage());
              }

              checkPoint++;
              clusterUpgradeJobService.setCheckPoint(clusterUpgradeJobId, node.getId(), checkPoint);

              int progress = (int) ((checkPoint * 100.0) / tasks.size());
              updateNodeProgress(node, progress);

              Thread.sleep(2000);
            } catch (Exception e) {
              node.setStatus(NodeUpgradeStatus.FAILED);
              int progress = (int) ((checkPoint * 100.0) / tasks.size());
              updateNodeProgress(node, progress);
              notifyNodeUpgradeFailed(node);

              log.error("Exception while running task [name: {}] for node [ip: {}]: {}", task.getName(), node.getIp(), e.getMessage(), e);
              throw new RuntimeException(e.getMessage(), e);
            } finally {
              notificationService.sendNotification(new UpgradeProgressChangeEvent());
            }
          }

          node.setVersion(config.targetVersion());
          node.setStatus(NodeUpgradeStatus.UPGRADED);
          updateNodeProgress(node, 100);
          notifyNodeUpgradedSuccessfully(cluster, node);
        }

      } catch (Exception e) {
        clusterUpgradeJobService.setJobStatus(clusterUpgradeJobId, ClusterUpgradeStatus.FAILED);
        if (nodes.size() > 1) {
          notifyClusterUpgradeFailed(cluster);
        }
        log.error("[ClusterId: {}] Cluster upgrade failed", cluster.getId(), e);
      } finally {
        clusterService.syncClusterState(cluster.getId());
        syncUpgradeJobStatus(cluster, clusterUpgradeJobId);
        notificationService.sendNotification(new UpgradeProgressChangeEvent());
        MDC.clear();
        clusterLockService.unlock(cluster.getId());
      }
    });
  }

  private void syncUpgradeJobStatus(SelfManagedClusterEntity cluster, String clusterUpgradeJobId) {
    final ClusterUpgradeJobEntity clusterUpgradeJob = clusterUpgradeJobService.getUpgradeJobById(clusterUpgradeJobId);
    List<GetClusterNodeResponse> nodes = clusterService.getNodes(cluster.getId(), null).stream()
        .filter(node -> node.status() != NodeUpgradeStatus.UPGRADED && !clusterUpgradeJob.getTargetVersion().equals(node.version()))
        .toList();
    if (nodes.isEmpty()) {
      clusterUpgradeJobService.setJobStatus(clusterUpgradeJob.getId(), ClusterUpgradeStatus.UPDATED);
      notifyClusterUpgradedSuccessFully(cluster);
    } else if (!clusterUpgradeJob.getStatus().equals(ClusterUpgradeStatus.STOPPED)) {
      var failedNodeExists = nodes
          .stream()
          .anyMatch(node -> NodeUpgradeStatus.FAILED.equals(node.status()));
      clusterUpgradeJobService.setJobStatus(clusterUpgradeJob.getId(),
          failedNodeExists ? ClusterUpgradeStatus.FAILED : ClusterUpgradeStatus.PARTIALLY_UPDATED
      );
    }
  }

  private void notifyClusterUpgradedSuccessFully(SelfManagedClusterEntity cluster) {
    String clusterName = cluster.getName();

    String message = String.format("Cluster '%s' has been successfully upgraded to the target version.", clusterName);
    String subject = String.format("Cluster '%s' upgraded", clusterName);

    notificationService.sendNotification(new GeneralNotificationEvent(NotificationType.SUCCESS, message, subject, cluster.getId()));
    notificationService.sendNotification(new UpgradeProgressChangeEvent());
  }

  private void notifyClusterUpgradedStopped(SelfManagedClusterEntity cluster) {
    String clusterName = cluster.getName();

    String message = String.format("Cluster '%s' has been successfully stopped.", clusterName);
    String subject = String.format("Cluster '%s' upgrade stoped", clusterName);

    notificationService.sendNotification(new GeneralNotificationEvent(NotificationType.SUCCESS, message, subject, cluster.getId()));
    notificationService.sendNotification(new UpgradeProgressChangeEvent());
  }

  private void notifyClusterUpgradeFailed(SelfManagedClusterEntity cluster) {
    String clusterName = cluster.getName();

    String message = String.format("Cluster '%s' failed to upgrade to the target version. Please investigate the issue.", clusterName);
    String subject = String.format("Cluster '%s' upgrade failed", clusterName);

    notificationService.sendNotification(new GeneralNotificationEvent(NotificationType.ERROR, message, subject, cluster.getId()));
    notificationService.sendNotification(new UpgradeProgressChangeEvent());
  }

  private void notifyNodeUpgradeFailed(ClusterNodeEntity node) {
    String message = String.format("Failed to upgrade node '%s' to the target version. Please check logs for details.", node.getName());

    notificationService.sendNotification(
        new GeneralNotificationEvent(NotificationType.ERROR, message, String.format("Node '%s' upgrade failed", node.getName()),
            node.getClusterId()));
    notificationService.sendNotification(new UpgradeProgressChangeEvent());
  }

  private void notifyNodeUpgradedSuccessfully(SelfManagedClusterEntity cluster, ClusterNodeEntity node) {
    notificationService.sendNotification(new UpgradeProgressChangeEvent());
    String message = String.format("Node '%s' has been successfully upgraded to the target version.", node.getName());
    notificationService.sendNotification(
        new GeneralNotificationEvent(NotificationType.SUCCESS, message, String.format("Node '%s' upgraded", node.getName()),
            cluster.getId()));
  }

  private void updateNodeProgress(ClusterNodeEntity node, int progress) {
    node.setProgress(progress);
    clusterNodeRepository.save(node);
    notificationService.sendNotification(new UpgradeProgressChangeEvent());
  }

}
